kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,248 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+#if !defined(KRATOS_PARALLEL_UTILITIES_H_INCLUDED )
+#define  KRATOS_PARALLEL_UTILITIES_H_INCLUDED
+
+
+// System includes
+#include<iostream>
+#include<array>
+#include<vector>
+#include<cmath>
+#include<limits>
+#include<omp.h>
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup KratosCore
+
+//***********************************************************************************
+//***********************************************************************************
+//***********************************************************************************
+template<class TContainerType, class TIteratorType=typename TContainerType::iterator>
+class BlockPartition
+{
+public:
+
+    static constexpr int max_allowed_threads = 128;
+
+    BlockPartition(TIteratorType it_begin,
+                   TIteratorType it_end,
+                   int Nchunks = omp_get_max_threads())
+        : mit_begin(it_begin), mit_end(it_end), mNchunks(Nchunks)
+    {
+        int NumTerms = it_end-it_begin;
+        int mBlockPartitionize = NumTerms / mNchunks;
+        mBlockPartition[0] = 0;
+        mBlockPartition[mNchunks] = NumTerms;
+        for(int i = 1; i < mNchunks; i++)
+            mBlockPartition[i] = mBlockPartition[i-1] + mBlockPartitionize ;
+
+    }
+
+    BlockPartition(TContainerType& rData,
+                   int Nchunks = omp_get_max_threads())
+        : BlockPartition(rData.begin(), rData.end(), Nchunks)
+    {}
+
+    virtual ~BlockPartition() {}
+
+    //version enabling reduction
+    template <class TFunctionWithReduction, class TReduceHelper>
+    inline void for_each(
+        TReduceHelper& reduce_helper,
+        TFunctionWithReduction&& f
+    )
+    {
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            TReduceHelper local_helper;
+            for (int k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                f(*(mit_begin + k), local_helper); //note that we pass a reference to the value, not the iterator
+            }
```

Comment:
```suggestion
            for (auto p = mit_begin + mBlockPartition[i], e = mit_begin + mBlockPartition[i+1]; p != e; ++p)
            {
                f(*p, local_helper); //note that we pass a reference to the value, not the iterator
            }
```
This should be the same with fewer arithmetics. Also, looks like you are only using `mBlockPartition` to get shifts into `mit_begin`. Why not store pointers in it directly instead of ints? Then the above loop could be
```cpp
            for (auto p = mBlockPartition[i]; p != mBlockPartition[i+1]; ++p)
            {
                f(*p, local_helper); //note that we pass a reference to the value, not the iterator
            }
```